src/computability/DFA.lean

```diff
@@ -128,3 +128,38 @@ begin
 end
 
 end DFA
+
+namespace language
+
+/-- A language `is_regular` if it is the strings accepted by some DFA -/
+structure is_regular (l : language α) :=
```

Comment:
This structure's fields contain data in `σ` and `M`, and we usually only name structures `is_X` when they are `Prop`s.

Also, you're using `is_regular l` hypotheses below in square brackets, and these are usually instantiated via type class search, so either those arguments should be made explicit, or this structure should be turned into a class.

Maybe you can say a little bit about how you envision using this structure, as the right design will depend on what you plan to do with it, and the usages below are all fairly trivial.

Reply:
I would guess that this probably _should_ be a `Prop`

Reply:
Indeed, `is_regular` should be a Prop.  (And it should have a single universe parameter.)  There are a lot of equivalent properties that describe regular languages (accepted by a DFA, matched by a regex, generated by a regular grammar, having finite index, ...).  The idea is that `is_regular` should provide an API that is independent of the concrete model.

Probably the easiest way around the universe issues is to define regularity in terms of regular expressions (though we're still missing the equivalence between regexps and DFAs for the pumping lemma then):
```lean
@[irreducible] def is_regular (l : language α) : Prop :=
∃ R : regular_expression α, l = R.matches
```

If we want to stick with DFAs for the definition, then `σ` should be in Type, as Eric commented:
```lean
@[irreducible] def is_regular (l : language α) : Prop :=
∃ (σ : Type) [fintype σ] (M :  DFA α σ), l = M.accepts
```

(Note that you can't use structure to define this type if it is in Prop, because then Lean can't create the projection to the fields `σ`, `_inst_1` and `M`.)

Reply:
I think `∃ (n : nat) (M :  DFA α (fin n)), l = M.accepts` might be slightly easier to work with, as it avoids dealing with instances in quantifiers

Reply:
You can then add a lemma to say that the universe-polymorphic statement is equivalent.